File: examples/python/destructive_guard.py
#!/usr/bin/env python3
"""Destructive Command Guard — Python version.

Blocks rm -rf on sensitive paths, git reset --hard, git clean -fd,
PowerShell Remove-Item, and sudo with dangerous commands.

Usage in settings.json:
{
    "hooks": {
        "PreToolUse": [{
            "matcher": "Bash",
            "hooks": [{"type": "command", "command": "python3 /path/to/destructive_guard.py"}]
        }]
    }
}
"""

import json
import re
import sys

def main():
    try:
        data = json.load(sys.stdin)
    except (json.JSONDecodeError, EOFError):
        sys.exit(0)

    command = data.get("tool_input", {}).get("command", "")
    if not command:
        sys.exit(0)

    # rm -rf on sensitive paths
    if re.search(r'rm\s+(-[rf]+\s+)*(\/|~|\.\.\/)', command):
        safe_dirs = {"node_modules", "dist", "build", ".cache", "__pycache__", "coverage"}
        if not any(command.rstrip().endswith(d) for d in safe_dirs):
            block("rm on sensitive path detected")

    # git reset --hard
    if re.search(r'(^|;|&&)\s*git\s+reset\s+--hard', command):
        block("git reset --hard discards all uncommitted changes")

    # git clean -fd
    if re.search(r'(^|;|&&)\s*git\s+clean\s+-[a-z]*[fd]', command):
        block("git clean removes untracked files permanently")

    # git checkout/switch --force
    if re.search(r'(^|;|&&)\s*git\s+(checkout|switch)\s+.*(--force|-f\b)', command):
        block("git checkout --force discards uncommitted changes")

    # PowerShell destructive
    if re.search(r'Remove-Item.*-Recurse.*-Force|rd\s+/s\s+/q', command, re.IGNORECASE):
        block("Destructive PowerShell command detected")

    # sudo with dangerous commands
    if re.search(r'^\s*sudo\s+(rm\s+-[rf]|chmod\s+(-R\s+)?777|dd\s+if=|mkfs)', command):
        block("sudo with dangerous command detected")

    sys.exit(0)

def block(reason):
    print(f"BLOCKED: {reason}", file=sys.stderr)
    sys.exit(2)

if __name__ == "__main__":
    main()

File: examples/rust/destructive_guard.rs
// destructive_guard.rs — Claude Code PreToolUse hook in Rust
//
// Blocks rm -rf /, git reset --hard, git clean -fd, and similar
// destructive commands. Exit code 2 = block, 0 = allow.
//
// Build: rustc destructive_guard.rs -o destructive-guard
// Usage: {"type": "command", "command": "/path/to/destructive-guard"}

use std::io::{self, Read};
use std::process;

fn main() {
    let mut input = String::new();
    if io::stdin().read_to_string(&mut input).is_err() {
        process::exit(0);
    }

    // Simple JSON parsing without serde (zero dependencies)
    let cmd = extract_command(&input);
    if cmd.is_empty() {
        process::exit(0);
    }

    // Skip echo/printf context
    let trimmed = cmd.trim_start().to_lowercase();
    if trimmed.starts_with("echo ") || trimmed.starts_with("printf ") {
        process::exit(0);
    }

    let patterns: &[&str] = &[
        "rm -rf /",
        "rm -rf ~/",
        "rm -rf ../",
        "rm -rf .",
        "git reset --hard",
        "git clean -f",
        "git checkout --force",
        "chmod 777 /",
        "find / -delete",
        "--no-preserve-root",
        "sudo mkfs",
    ];

    for pattern in patterns {
        if cmd.to_lowercase().contains(&pattern.to_lowercase()) {
            eprintln!("BLOCKED: Dangerous command detected");
            eprintln!("Command: {}", cmd);
            process::exit(2);
        }
    }

    process::exit(0);
}

fn extract_command(json: &str) -> String {
    // Extract .tool_input.command from JSON without a parser
    if let Some(pos) = json.find("\"command\"") {
        let rest = &json[pos + 9..];
        if let Some(start) = rest.find('"') {
            let value_start = start + 1;
            let mut end = value_start;
            let bytes = rest.as_bytes();
            while end < bytes.len() {
                if bytes[end] == b'"' && (end == 0 || bytes[end - 1] != b'\\') {
                    return rest[value_start..end].replace("\\\"", "\"").replace("\\\\", "\\");
                }
                end += 1;
            }
        }
    }
    String::new()
}
